### PR TITLE
[AIRFLOW-6005] Common classes should always be imported from airflow, Depends on [AIRFLOW-6128]

### PR DESCRIPTION
**Pull request by potiuk**
NOTE FOR REVIEWER: This change depends on #6596 #6702 #6696  so please review only the last commit.

For the sake of consistency, all the common components exposed directly
in 'airflow' __init__.py should be imported from there rather than
from sub-packages. This is important also for back-porting
purposes (AIP-21) to make sure that we are using stable imports.

This change does it and implements pre-commit that prevents importing
those common imports differently.

Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-6005

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

**Comment by potiuk** [comment]
I think I nicely solved the problem of old imports. I figured out a nice way how to throw a deprecation warning in case someone import "from airflow impot AirflowException". It will still work, but during tests/execution it will log a deprecation warning and also you will see this deprecation warning in IDE:

Screenshot 2019-11-18 at 02 38 00

This is a I think a very straightforward way to avoid potential cyclic imports like the ones described in #6596 and opens up the possibility of removing the class from 'airflow' package in the future.

I have one problem though @mik-laj maybe you can help solving it. I added a new "deprecations" package where i will put all the "deprecated" classes. I had to remove it in `autoapi_ignore` because otherwise there were two candidates for AirflowException class in autoapi generation. Now I have another warning - when I generate doc I have the "missing module" error:

```
[AutoAPI] Mapping Data
WARNING: Cannot resolve import of unknown module airflow.deprecations in airflow
[AutoAPI] Rendering Data
```
I tried to add `airflow.deprecations` to 'autodoc_mock_imports' but it did not help. Can you please help with solving this?

**Comment by potiuk** [comment]
I think I found a nice (though a but hackish :) ) way of making the docs build @mik-laj and others. See the change :).

File: airflow/deprecations/airflow_exception_deprecated.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import warnings
+
+# noinspection PyUnresolvedReferences
+from airflow.exceptions import AirflowException as AirflowExceptionsDeprecated
+
+
+class AirflowException(AirflowExceptionsDeprecated):
+    """Deprecated version of AirflowException"""
+    def __init__(self, *args):
+        warnings.warn(
```
Comment (ashb) [inline airflow/deprecations/airflow_exception_deprecated.py]: This is going to warn at throw time, only right? What is the pattern we used for AIP-21?

And as with you rother PR, for small classes I would just put this inside airflow/deprecations/__init__.py (or airflow/deprecations.py) -- each extra file we have to import adds a small penalty to Airflow startup time (which is already noticeable.)

File: airflow/__init__.py [hunk 2]
```diff
@@ -29,15 +29,20 @@
...
 
 from airflow import settings
 from airflow import version
 from airflow.utils.log.logging_mixin import LoggingMixin
 from airflow.configuration import conf
-from airflow.exceptions import AirflowException
 from airflow.models import DAG
 
+if 'BUILDING_AIRFLOW_DOCS' in os.environ:
+    from airflow.exceptions import AirflowException
```
Comment (ashb) [inline airflow/__init__.py]: Why do we need this when building docs?

**Review by ashb — CHANGES_REQUESTED** [review]

File: airflow/deprecations/airflow_exception_deprecated.py (hunk 1 above)
Comment (potiuk) [reply airflow/deprecations/airflow_exception_deprecated.py]: Yep. that was an experiment - I tried to make docs build. And tried different approaches but originally I had It in the _init__ and I am going to put it back.

**Comment by ashb** [comment]
Oh, I think I see the issue with both of my comments:

Because `airflow/__init__.py` unconditionally imports `from airflow.deprecations import AirflowException` it would always run.

~Could we use https://docs.python.org/3/library/importlib.html#importlib.util.LazyLoader (new in v3.5) instead?~ That doesn't seem to work for an `pip install -e` environment so no. https://pypi.org/project/lazy-import/ might be an option though. Oh, this module is GPLv3 so won't work.

And as for the "two candidates for AirflowException class in autoapi generation": putting `__all__ = ['AirflowException']` in the airflow/deprecations/airflow_exception_deprecated.py might help with that.

File: airflow/__init__.py (hunk 2 above)
Comment (potiuk) [reply airflow/__init__.py]: The problem is explained above. when I had 'deprecated' AirflowEccwption I had errro about 'duplicate candidates' for autoapi. So I had to exclude the whole 'deprecations' folder from autoapi - but this caused a problem of 'missing import module' when generating docs. I tried to mock it using 'autodoc_mock_imports' but it did not work for airflow internal modules. So I found that I can use the same mechanism as used in 'apply_defaults' with the variable.

**Comment by potiuk** [comment]
@ashb - I will try those approaches. First I have to fix the #6596 which is the base for that one and then will try some experimenting.

File: airflow/__init__.py (hunk 2 above)
Comment (potiuk) [reply airflow/__init__.py]: Indeed getting __all__ solved the problem. Thanks @ashb

**Comment by potiuk** [comment]
@ashb - I looked at the options and now I implemented quickly version that does not do deprecation, but pre-commits. It requires:

- 'from airflow import AirflowException' in "providers" and "contrib"
- 'from airflow.exceptions import AirflowException' elsewhere

I think that might make perfect sense rather than deprecating. It's even more "constraining" for core developers. The check is pretty much self-explanatory:

```
[potiuk:/code/airflow] remove-airflow-airflow-exception-imports+ 3d11h32m58s ± pre-commit run airflow-exception --all-files
Make sure AirflowException is imported from airflow.exceptions in core.......................Passed
Make sure AirflowException is imported from airflow outside of core..........................Passed
[potiuk:/code/airflow] remove-airflow-airflow-exception-imports+ 3d11h33m25s ± pre-commit run airflow-settings --all-files
Make sure settings are imported individually as local imports.......................Passed
```

**Comment by potiuk** [comment]
@ashb @mik-laj @feluelle -> That's the next one resulting from the untangling cyclic deps. This one should be helpful in keeping consistent import for common classes inside airflow codebase.

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -237,6 +237,43 @@ repos:
...
+        exclude: >
+          (?x)
+          ^airflow/__init__\.py$|
+          ^airflow/logging_config.py$|
+          ^airflow/settings.py$
+        files: \.py$
+      - id: common-imports
+        language: pygrep
+        name: Make sure AirflowConfigException is imported consistently 'from airflow'
+        entry: "from airflow\\.exception import.* AirflowConfigException"
```
Comment (feluelle) [inline .pre-commit-config.yaml:266]:
```suggestion
        entry: "from airflow\\.exceptions import.* AirflowConfigException"
```

**Review by feluelle — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml (hunk 3 above)
Comment (potiuk) [reply .pre-commit-config.yaml:266]: Will come back to it some day :).

**Comment by potiuk** [comment]
Since we are going to get rid of the airflow.__init__.py most likely - this one can be closed.

**Pull request closed without merging**